### PR TITLE
Increase timeout of intproxy connection to a ludicrous amount (avoids intproxy issues when debugging with breakpoints).

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: #2652 
#2750 (default intproxy logging catching a ride)

I couldn't really reproduce the issue, but this should be enough to avoid it.

### CODE REVIEW

File: mirrord/cli/src/internal_proxy.rs [hunk 1]
```diff
@@ -53,25 +56,42 @@ pub(crate) async fn proxy(watch: drain::Watch) -> Result<(), InternalProxyError>
...
+        .log_destination
+        .as_ref()
+        .map(PathBuf::from)
+        .unwrap_or_else(|| {
+            let random_name: String = rand::thread_rng()
+                .sample_iter(&Alphanumeric)
+                .take(7)
+                .map(char::from)
+                .collect();
+            let timestamp = Utc::now().timestamp();
```
Comment (Razz4780) [inline mirrord/cli/src/internal_proxy.rs:71]:
```suggestion
            let timestamp = Utc::now().timestamp();
```
```suggestion
            let timestamp = SystemTime::UNIX_EPOCH.elapsed().unwrap();
```
Can be done with Rust std, no need to add `chrono` :p

File: mirrord/cli/src/internal_proxy.rs [hunk 2]
```diff
@@ -53,25 +56,42 @@ pub(crate) async fn proxy(watch: drain::Watch) -> Result<(), InternalProxyError>
...
+        .open(&log_destination)
+        .map_err(|fail| {
+            InternalProxyError::OpenLogFile(log_destination.to_string_lossy().to_string(), fail)
+        })?;
+
+    let log_level = config
+        .internal_proxy
+        .log_level
+        .clone()
+        .unwrap_or_else(|| "mirrord=trace".to_string());
```
Comment (Razz4780) [inline mirrord/cli/src/internal_proxy.rs:88]:
```suggestion
        .clone()
        .unwrap_or_else(|| "mirrord=trace".to_string());
```
```suggestion
        .as_deref()
        .unwrap_or("mirrord=trace");
```
We don't need owned, I think

File: mirrord/config/src/internal_proxy.rs [hunk 3]
```diff
@@ -84,6 +84,22 @@ pub struct InternalProxyConfig {
...
+    /// Sometimes the cpu is too busy with other tasks and the internal proxy is blocked
+    /// from performing its keep-alive check. Increase this value if mirrord exits while
+    /// you're debugging with breakpoints.
+    ///
+    /// ```json
+    /// {
+    ///   "internal_proxy": {
+    ///     "idle_timeout": 31536000
+    ///   }
+    /// }
```
Comment (Razz4780) [inline mirrord/config/src/internal_proxy.rs:98]: This is kind of misleading :v
1. The default value is already at max, so "Increase this value if mirrord exits while you're debugging with breakpoints" does not sound right
2. Snippet contains `idle_timeout` field
3. The explanation itself is kinda off, the problem is not about keep-alive checks

This setting is not really for internal proxy but for the layer, since we set the timeout there. I propose we move this into `experimental` section under `.experimental.intproxy_socket_timeout` without any instructions (just a doc saying that it's a read/write timeout for layer<->intproxy sockets, the users should not normally touch it without our directions)

File: mirrord/cli/src/internal_proxy.rs (hunk 2 above)
Comment (Razz4780) [inline mirrord/cli/src/internal_proxy.rs:88]: Not sure if we should default to `trace`. That's gonna produce large files for long mirrord sessions.
Maybe we should:
1. Move default logging change to separate PR
2. Use `mirrord=info` as default level
3. Make sure that intproxy logs on `info` are not spammy and still informative (only layer connects/disconnects and files/connections/subscriptions state changes)
 @aviramha wdyt?

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: mirrord/config/src/internal_proxy.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/config/src/internal_proxy.rs:98]: I don't think the user should be seeing this either, but it feels like it belong under the `internal_proxy` config. I mean, this is a setting for an internal proxy startup thingy. Aren't the `experimental` configs for things we're not quite sure we want? This one feels like it doesn't hurt to have, so I see no reason to remove it in the future.

Comment (Razz4780) [reply mirrord/config/src/internal_proxy.rs]: Tbh I think `experimental` is mainly for hacks :V

> this is a setting for an internal proxy startup thingy

Wait, no. You're thinking about `.internal_proxy.idle_timeout`. My comment is for your new setting. The timeout specified in it is set on layer's TCP socket for communicating with intproxy (so applies to any request<->response exchange)

Comment (meowjesty) [reply mirrord/config/src/internal_proxy.rs]: > layer's TCP socket for communicating with intproxy

Yeah I got that, but isn't that part of the proxy startup? When we have a new layer connection? That's what I mean.

Comment (meowjesty) [reply mirrord/config/src/internal_proxy.rs]: Anyway, I changed the docs.

File: mirrord/cli/src/internal_proxy.rs (hunk 2 above)
Comment (meowjesty) [reply mirrord/cli/src/internal_proxy.rs]: New PR just for this at #2764

Comment (aviramha) [reply mirrord/cli/src/internal_proxy.rs]: sounds like a plan.

File: mirrord/layer/src/lib.rs [hunk 4]
```diff
@@ -164,7 +164,7 @@ static EXECUTABLE_PATH: OnceLock<String> = OnceLock::new();
 
 /// Proxy Connection timeout
 /// Set to 10 seconds as most agent operations timeout after 5 seconds
```
Comment (Razz4780) [inline mirrord/layer/src/lib.rs:166]:
```suggestion
/// Read/write timeout for layer<->intproxy TCP sockets.
/// Can be configured in the [`LayerConfig`].
```

File: mirrord/cli/Cargo.toml [hunk 5]
```diff
@@ -68,6 +68,8 @@ tokio-stream = { workspace = true, features = ["net"] }
 tokio-retry = "0.3"
 regex.workspace = true
 mid = "2.1.0"
+rand.workspace = true
+chrono = { version = "0.4" }
```
Comment (Razz4780) [inline mirrord/cli/Cargo.toml:72]: Delet :wastebasket:

File: mirrord/cli/src/internal_proxy.rs [hunk 6]
```diff
@@ -53,25 +55,42 @@ pub(crate) async fn proxy(watch: drain::Watch) -> Result<(), InternalProxyError>
...
+        .internal_proxy
+        .log_level
+        .as_deref()
+        .unwrap_or("mirrord=trace");
+
+    tracing_subscriber::fmt()
+        .with_writer(output_file)
+        .with_ansi(false)
+        .with_env_filter(EnvFilter::builder().parse_lossy(log_level))
+        .init();
```
Comment (Razz4780) [inline mirrord/cli/src/internal_proxy.rs:93]: Delet :wastebasket:

File: mirrord/layer/src/lib.rs [hunk 7]
```diff
@@ -626,7 +630,10 @@ pub(crate) unsafe extern "C" fn fork_detour() -> pid_t {
             let new_connection = ProxyConnection::new(
                 parent_connection.proxy_addr(),
                 NewSessionRequest::Forked(parent_connection.layer_id()),
-                PROXY_CONNECTION_TIMEOUT,
+                PROXY_CONNECTION_TIMEOUT
+                    .get()
+                    .copied()
+                    .unwrap_or_else(|| Duration::from_secs(31536000)),
```
Comment (Razz4780) [inline mirrord/layer/src/lib.rs:636]: Should be initialized, I'd just do `.get().expect(...)`, so we don't need this value hardcoded here

File: mirrord-schema.json [hunk 8]
```diff
@@ -1247,6 +1247,16 @@
       "description": "Configuration for the internal proxy mirrord spawns for each local mirrord session that local layers use to connect to the remote agent\n\nThis is seldom used, but if you get `ConnectionRefused` errors, you might want to increase the timeouts a bit.\n\n```json { \"internal_proxy\": { \"start_idle_timeout\": 30, \"idle_timeout\": 5 } } ```",
       "type": "object",
       "properties": {
+        "breakpoint_timeout": {
```
Comment (Razz4780) [inline mirrord-schema.json:1250]: Needs regeneration

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: mirrord/config/src/internal_proxy.rs (hunk 3 above)
Comment (Razz4780) [reply mirrord/config/src/internal_proxy.rs]: Oke, let's leave it in `.internal_proxy` section

**Review by Razz4780 — APPROVED** [review]
